File: app/src/main/java/forpdateam/ru/forpda/presentation/history/HistoryPresenter.java
package forpdateam.ru.forpda.presentation.history;

import android.os.Bundle;

import com.arellomobile.mvp.InjectViewState;

import forpdateam.ru.forpda.common.IntentHandler;
import forpdateam.ru.forpda.common.Utils;
import forpdateam.ru.forpda.common.mvp.BasePresenter;
import forpdateam.ru.forpda.entity.app.history.HistoryItem;
import forpdateam.ru.forpda.model.repository.HistoryRepository;
import forpdateam.ru.forpda.ui.fragments.TabFragment;
import io.reactivex.disposables.Disposable;

/**
 * Created by radiationx on 11.11.17.
 */

@InjectViewState
public class HistoryPresenter extends BasePresenter<HistoryView> {

    private HistoryRepository historyRepository;

    public HistoryPresenter(HistoryRepository historyRepository) {
        this.historyRepository = historyRepository;
    }

    public void getHistory() {
        Disposable disposable
                = historyRepository.getHistory()
                .doOnTerminate(() -> getViewState().setRefreshing(true))
                .doAfterTerminate(() -> getViewState().setRefreshing(false))
                .subscribe(historyItemBds -> {
                    getViewState().showHistory(historyItemBds);
                }, this::handleErrorRx);
        addToDisposable(disposable);
    }

    public void remove(int id) {
        Disposable disposable
                = historyRepository.remove(id)
                .doOnTerminate(() -> getViewState().setRefreshing(true))
                .doAfterTerminate(() -> getViewState().setRefreshing(false))
                .subscribe(this::getHistory, this::handleErrorRx);
        addToDisposable(disposable);
    }

    public void clear() {
        Disposable disposable
                = historyRepository.clear()
                .doOnTerminate(() -> getViewState().setRefreshing(true))
                .doAfterTerminate(() -> getViewState().setRefreshing(false))
                .subscribe(this::getHistory, this::handleErrorRx);
        addToDisposable(disposable);
    }

    public void copyLink(HistoryItem item) {
        Utils.copyToClipBoard(item.getUrl());
    }

    public void onItemClick(HistoryItem item) {
        Bundle args = new Bundle();
        args.putString(TabFragment.ARG_TITLE, item.getTitle());
        IntentHandler.handle(item.getUrl(), args);
    }

    public void onItemLongClick(HistoryItem item) {
        getViewState().showItemDialogMenu(item);
    }
}

File: app/src/main/java/forpdateam/ru/forpda/presentation/history/HistoryPresenter.kt
package forpdateam.ru.forpda.presentation.history

import forpdateam.ru.forpda.common.Utils
import forpdateam.ru.forpda.common.mvp.BasePresenter
import forpdateam.ru.forpda.entity.app.history.HistoryItem
import forpdateam.ru.forpda.extensions.coRunCatching
import forpdateam.ru.forpda.model.repository.history.HistoryRepository
import forpdateam.ru.forpda.presentation.ErrorHandler
import forpdateam.ru.forpda.presentation.LinkHandler
import forpdateam.ru.forpda.presentation.Screen
import forpdateam.ru.forpda.presentation.TabRouter
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.flow.onStart
import kotlinx.coroutines.launch
import moxy.InjectViewState

/**
 * Created by radiationx on 11.11.17.
 */

@InjectViewState
class HistoryPresenter(
    private val historyRepository: HistoryRepository,
    private val router: TabRouter,
    private val linkHandler: LinkHandler,
    private val errorHandler: ErrorHandler,
    private val utils: Utils
) : BasePresenter<HistoryView>() {

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        historyRepository
            .observeItems()
            .onStart {
                viewState.setRefreshing(true)
            }
            .onEach {
                viewState.setRefreshing(false)
                viewState.showHistory(it)
            }
            .launchIn(viewModelScope)
    }


    fun remove(id: Int) {
        viewModelScope.launch {
            coRunCatching {
                historyRepository.remove(id)
            }.onFailure {
                errorHandler.handle(it)
            }
        }
    }

    fun clear() {
        viewModelScope.launch {
            coRunCatching {
                historyRepository.clear()
            }.onFailure {
                errorHandler.handle(it)
            }
        }
    }

    fun copyLink(item: HistoryItem) {
        utils.copyToClipBoard(item.url)
    }

    fun onItemClick(item: HistoryItem) {
        linkHandler.handle(
            item.url, mapOf(
                Screen.ARG_TITLE to item.title
            )
        )
    }

    fun onItemLongClick(item: HistoryItem) {
        viewState.showItemDialogMenu(item)
    }
}
